corehq/apps/accounting/forms.py

```diff
@@ -1944,10 +1944,13 @@ class EnterprisePlanContactForm(forms.Form):
     )
 
     def __init__(self, domain, web_user, data=None, *args, **kwargs):
+        from corehq.apps.domain.views.accounting import SelectPlanView
+        back_button_text = kwargs.pop('back_button_text', 'Select different plan')
+        back_button_url = kwargs.pop('back_button_url', SelectPlanView.urlname)
```

Comment:
What do you think about using declared parameters rather than `kwargs` here? specifically, we could change this to be:
`def __init__(self, domain, web_user, back_button_text=gettext_lazy('Select different plan'), back_button_url=None, data=None, *args, **kwargs)`, or you could use the named the args syntax like:
`def __init__(self, domain, web_user, data=None, *args, back_button_text=gettext_lazy('Select different plan'), back_button_url=None, **kwargs)`. 

I'm also curious about the need for defaults here. I assumed we provided defaults because multiple forms re-used the same values, but I notice that both `AnnualPlanContactForm` and `CustomPlanContactForm` use `back_button_text` of "Back to my Subscription", while `EnterprisePlanContactForm` relies on the default value. I wonder if it would be more straight-forward to remove the default and have each class provide its own value.

Finally, I think `back_button_url` is confusing with how it's currently used, but conveys the right idea. Currently, it's not actually a URL, but a "URL name" that is resolved later. Could we instead ask the caller to pass in a fully-formed URL? That would let future callers, who use urls that need more or less than a domain argument, to still use this class.

Reply:
`AnnualPlanContactForm` conditionally uses "Back to my Subscription" but falls back to the "Select different plan" default if not `on_annual_plan` -- which is to say if someone is trying to sign up for an annual plan for the first time, because they would have gotten there from the "Select Plan" page. Since there are two uses of each version, it is probably equally valid to have either of them as the default, or to have no explicit default set.

I do think the idea of passing in the url itself might make sense here. Would you do that from the form (keeping some local imports) or from the view? I don't love the idea of the view having to specify which buttons we're showing on the form, but I'm not sure how that balances against local imports.

Reply:
So, I'm trying to understand what `AnnualPlanContactForm` wants to represent -- what generic concept. I think this is meant to be a form that displays `name`, `company`, and `message` fields and has configurable back and submit buttons that adhere to standardized styling. Additionally, it needs to be able to send an email with a custom title. However, I don't think this is specific to `AnnualPlanContactForm` -- the above description also describes `EnterprisePlanContactForm` and `CustomPlanContactForm`. Given that each of these forms is only used by their respective view, I think this might be a case of redundant classes. My preference would be to make the Form itself generic, which you've already done with `PlanContactForm`, while removing `EnterprisePlanContactForm`, `AnnualPlanContactForm`, and `CustomPlanContactForm`. I'd have the constructor take in properties for `back_button_text` (default to something basic like 'Back') and `back_button_url`. And I'd modify `send_message` to take in a `request_type` parameter. Then each View can use `PlanContactForm`, customizing it with the back button's text and URL, and providing its own request text when calling `send_message`. 

That said, I mentioned these were non-blocking suggestions. I'll defer to your best judgment on what you think makes the most sense.

Reply:
I waffled on this myself a bit (how much refactoring is _too_ much refactoring), but ultimately I think you're right on the class duplication. The subclasses aren't really doing anything here when the form is truly identical, sending 99% the same message to the same place.